### PR TITLE
Resolve partial update conflict

### PR DESCRIPTION
**Pull request by sevev**
Check and resolve conflict in RowsetUpdateStates::apply if a conflict exists.
1. If the row has been upsert, reread this row's value
2. If the row has been deleted, set this row's value to its default value

### CODE REVIEW

File: be/src/storage/rowset_update_state.cpp [hunk 1]
```diff
@@ -247,20 +247,103 @@ Status RowsetUpdateState::_prepare_partial_update_states(Tablet* tablet, Rowset*
...
+
+            for (size_t col_idx = 0; col_idx < read_column_ids.size(); col_idx++) {
+                new_write_columns[col_idx]->append_selective(*read_columns[col_idx], read_idxes.data(), 0,
+                                                             read_idxes.size());
+            }
+
+            for (size_t col_idx = 0; col_idx < _partial_update_states[i].write_columns.size(); col_idx++) {
+                RETURN_IF_ERROR(_partial_update_states[i].write_columns[col_idx]->update_rows(
+                        *new_write_columns[col_idx], conflict_idxes.data()));
+            }
```
Comment (decster) [inline be/src/storage/rowset_update_state.cpp:332]: These 2 loops could be combined into one, and create new_write_columns inside this loop?

File: be/src/storage/rowset_update_state.cpp [hunk 2]
```diff
@@ -247,20 +247,103 @@ Status RowsetUpdateState::_prepare_partial_update_states(Tablet* tablet, Rowset*
...
+        }
+        int64_t t_resolve_conflict_end = MonotonicMillis();
+        LOG(INFO) << Substitute(
+                "check partial rowset conflict tablet:$0 rowset:$1 seg:$2 #column:$3 #conflict_rows:$4 getrowid:$5ms "
+                "#resolve_conflict $6ms",
+                tablet->tablet_id(), rowset_id, i, read_column_ids.size(), conflict_idxes.size(),
+                (t_get_rowids - t_start) / 1000000, (t_resolve_conflict_end - t_resolve_conflict_start) / 1000000);
```
Comment (decster) [inline be/src/storage/rowset_update_state.cpp:339]: t_get_rowids - t_start  are all the same, maybe we can gather all segments' information and print log after all segments are processed(out of the loop)

**Review by decster — CHANGES_REQUESTED** [review]

File: be/src/storage/rowset_update_state.cpp (hunk 1 above)
Comment (sevev) [reply be/src/storage/rowset_update_state.cpp]: > These 2 loops could be combined into one, and create new_write_columns inside this loop?

done，change in https://github.com/StarRocks/starrocks/pull/2919/commits/0d57187a8d7ba3a80d526c346f5ff7d10704559d

File: be/src/storage/rowset_update_state.cpp (hunk 2 above)
Comment (sevev) [reply be/src/storage/rowset_update_state.cpp]: > t_get_rowids - t_start are all the same, maybe we can gather all segments' information and print log after all segments are processed(out of the loop)

done, change in https://github.com/StarRocks/starrocks/pull/2919/commits/0d57187a8d7ba3a80d526c346f5ff7d10704559d

**Comment by sevev** [comment]
run starrocks_be_unittest

File: be/src/storage/rowset_update_state.cpp [hunk 3]
```diff
@@ -247,20 +247,101 @@ Status RowsetUpdateState::_prepare_partial_update_states(Tablet* tablet, Rowset*
...
+            size_t num_default = 0;
+            std::map<uint32_t, std::vector<uint32_t>> rowids_by_rssid;
+            std::vector<uint32_t> read_idxes;
+            plan_read_by_rssid(conflict_rowids, &num_default, &rowids_by_rssid, &read_idxes);
+            DCHECK_EQ(conflict_idxes.size(), read_idxes.size());
+            RETURN_IF_ERROR(tablet->updates()->get_column_values(read_column_ids, num_default > 0, rowids_by_rssid,
+                                                                 &read_columns));
+
+            for (size_t col_idx = 0; col_idx < read_column_ids.size(); col_idx++) {
+                new_write_columns[col_idx]->append_selective(*read_columns[col_idx], read_idxes.data(), 0,
```
Comment (decster) [inline be/src/storage/rowset_update_state.cpp:326]: new_write_columns is a temp var, which can be created in the loop

Comment (sevev) [reply be/src/storage/rowset_update_state.cpp]: > new_write_columns is a temp var, which can be created in the loop

done, https://github.com/StarRocks/starrocks/pull/2919/commits/5cd750b4c71639121b2c917ecabe38397a5adcab

**Review by decster — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**